File: forms/international-patient-summary/front-end-form-with-html/js/ips-validator.js
// IPS Completeness Validator. Pure functions: take an
// `AssessmentData` object, return the completeness level (Complete /
// Partial / Incomplete) plus per-section audit trail.
//
// Grading rules (per ISO 27269 / HL7 FHIR IPS IG):
//   All 8 mandatory + all 2 optional populated   -> Complete
//   All 8 mandatory populated, optional missing  -> Partial
//   Any 1 mandatory missing                      -> Incomplete

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').CompletenessLevel} CompletenessLevel
 * @typedef {import('./types.js').FiredRule} FiredRule
 */

// Wrapped in an IIFE; published via window.InternationalPatientSummary.
(function () {
'use strict';
window.InternationalPatientSummary = window.InternationalPatientSummary || {};
const { validationRules } = window.InternationalPatientSummary;

/**
 * @param {{ mandatoryPopulated: number, mandatoryTotal: number,
 *           optionalPopulated: number, optionalTotal: number }} counts
 * @returns {CompletenessLevel}
 */
function classifyCompleteness(counts) {
  if (counts.mandatoryPopulated < counts.mandatoryTotal) return 'incomplete';
  if (counts.optionalPopulated < counts.optionalTotal) return 'partial';
  return 'complete';
}

/**
 * Friendly label for a CompletenessLevel.
 * @param {CompletenessLevel} level
 */
function completenessLevelLabel(level) {
  switch (level) {
    case 'complete':   return 'Complete';
    case 'partial':    return 'Partial';
    case 'incomplete': return 'Incomplete';
    default:           return '';
  }
}

/**
 * CSS class hint for the completeness badge.
 * @param {CompletenessLevel} level
 */
function completenessLevelClass(level) {
  switch (level) {
    case 'complete':   return 'completeness-complete';
    case 'partial':    return 'completeness-partial';
    case 'incomplete': return 'completeness-incomplete';
    default:           return '';
  }
}

/**
 * Evaluate the IPS completeness rules against the supplied assessment
 * data and produce the completeness level plus per-section audit trail.
 *
 * @param {AssessmentData} data
 * @returns {{ completenessLevel: CompletenessLevel,
 *             mandatoryPopulated: number, mandatoryTotal: number,
 *             optionalPopulated: number, optionalTotal: number,
 *             firedRules: FiredRule[] }}
 */
function validateIPS(data) {
  /** @type {FiredRule[]} */
  const firedRules = [];
  let mandatoryPopulated = 0;
  let mandatoryTotal = 0;
  let optionalPopulated = 0;
  let optionalTotal = 0;

  for (const rule of validationRules) {
    let outcome = 'empty';
    try {
      outcome = rule.evaluate(data);
    } catch (e) {
      console.warn(`IPS rule ${rule.id} evaluation failed:`, e);
    }
    if (rule.mandatory) {
      mandatoryTotal++;
      if (outcome === 'ok') mandatoryPopulated++;
    } else {
      optionalTotal++;
      if (outcome === 'ok') optionalPopulated++;
      // Optional sections are surfaced as "optional" in the audit
      // table when empty so they read as informational rather than as
      // blocking failures.
      if (outcome === 'empty') outcome = 'optional';
    }
    firedRules.push({
      id: rule.id,
      category: rule.category,
      description: rule.description,
      status: outcome,
      mandatory: rule.mandatory
    });
  }

  const completenessLevel = classifyCompleteness({
    mandatoryPopulated,
    mandatoryTotal,
    optionalPopulated,
    optionalTotal
  });

  return {
    completenessLevel,
    mandatoryPopulated,
    mandatoryTotal,
    optionalPopulated,
    optionalTotal,
    firedRules
  };
}

Object.assign(window.InternationalPatientSummary, {
  classifyCompleteness,
  completenessLevelLabel,
  completenessLevelClass,
  validateIPS
});
})();

File: forms/international-patient-summary/back-end-with-loco/src/engine/ips_validator.rs
//! Ips validator module.

use super::flagged_issues::detect_additional_flags;
use super::types::{
    AssessmentData, CompletenessLevel, FiredRule, GradingResult,
};
use super::validation_rules::all_rules;

/// Classify the overall IPS completeness given populated / total counts.
///
/// - All mandatory + all optional populated -> `complete`
/// - All mandatory populated, optional missing -> `partial`
/// - Any 1 mandatory missing -> `incomplete`
pub fn classify_completeness(
    mandatory_populated: u32,
    mandatory_total: u32,
    optional_populated: u32,
    optional_total: u32,
) -> CompletenessLevel {
    if mandatory_populated < mandatory_total {
        return "incomplete".to_string();
    }
    if optional_populated < optional_total {
        return "partial".to_string();
    }
    "complete".to_string()
}

/// Friendly label for a completeness level.
pub fn completeness_level_label(level: &str) -> String {
    match level {
        "complete" => "Complete".to_string(),
        "partial" => "Partial".to_string(),
        "incomplete" => "Incomplete".to_string(),
        _ => String::new(),
    }
}

/// Pure function: evaluate the IPS completeness rules against the supplied
/// data and produce the completeness level plus per-section audit trail
/// and clinician-facing flags.
pub fn validate(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let mut fired_rules: Vec<FiredRule> = Vec::new();
    let mut mandatory_populated: u32 = 0;
    let mut mandatory_total: u32 = 0;
    let mut optional_populated: u32 = 0;
    let mut optional_total: u32 = 0;

    for rule in all_rules() {
        let mut outcome = (rule.evaluate)(data).to_string();
        if rule.mandatory {
            mandatory_total += 1;
            if outcome == "ok" {
                mandatory_populated += 1;
            }
        } else {
            optional_total += 1;
            if outcome == "ok" {
                optional_populated += 1;
            }
            // Optional sections surface as "optional" when empty so they
            // read as informational rather than blocking failures.
            if outcome == "empty" {
                outcome = "optional".to_string();
            }
        }
        fired_rules.push(FiredRule {
            id: rule.id.to_string(),
            category: rule.category.to_string(),
            description: rule.description.to_string(),
            status: outcome,
            mandatory: rule.mandatory,
        });
    }

    let completeness_level = classify_completeness(
        mandatory_populated,
        mandatory_total,
        optional_populated,
        optional_total,
    );
    let additional_flags = detect_additional_flags(data);

    GradingResult {
        completeness_level,
        mandatory_populated,
        mandatory_total,
        optional_populated,
        optional_total,
        fired_rules,
        additional_flags,
        timestamp,
    }
}
